cpu/cc2538/periph/uart.c

```diff
@@ -187,20 +207,71 @@ int uart_mode(uart_t uart, uart_data_bits_t data_bits, uart_parity_t parity,
 }
 #endif
 
+static inline void send_byte(uart_t uart, uint8_t byte)
+{
+    /* Block if the TX FIFO is full */
+    while (uart_config[uart].dev->cc2538_uart_fr.FRbits.TXFF) {}
+    uart_config[uart].dev->DR = byte;
+}
+
 void uart_write(uart_t uart, const uint8_t *data, size_t len)
 {
     assert(uart < UART_NUMOF);
-
-    cc2538_uart_t *u = uart_config[uart].dev;
-
-    /* Block if the TX FIFO is full */
+#ifdef MODULE_PERIPH_UART_NONBLOCKING
     for (size_t i = 0; i < len; i++) {
-        while (u->cc2538_uart_fr.FRbits.TXFF) {}
-        u->DR = data[i];
+        uart_config[uart].dev->cc2538_uart_im.IM |= TXMIS;
+        if (irq_is_in() || __get_PRIMASK()) {
+            /* if ring buffer is full free up a spot */
+            if (tsrb_full(&uart_tx_rb[uart])) {
+                send_byte(uart, tsrb_get_one(&uart_tx_rb[uart]));
+            }
+            /* if FIFO is full write to the buffer */
+            if (uart_config[uart].dev->cc2538_uart_fr.FRbits.TXFF) {
+                tsrb_add_one(&uart_tx_rb[uart], data[i]);
+            }
+            /* if tx FIFO is not full then add data to the FIFO */
+            else {
+                /* if there is already data in the tsrb buffer write that byte */
+                int byte = tsrb_get_one(&uart_tx_rb[uart]);
+                if (byte >= 0) {
+                    send_byte(uart, byte);
+                    tsrb_add_one(&uart_tx_rb[uart], data[i]);
+                }
+                /* If there is not data in the buffer directly write the
+                   current byte*/
+                else {
+                    send_byte(uart, data[i]);
+                }
+            }
+        }
+        else {
+            /* If Tx FIFO is full then add to ring buffer */
+            if (uart_config[uart].dev->cc2538_uart_fr.FRbits.TXFF) {
+                while (tsrb_add_one(&uart_tx_rb[uart], data[i]) < 0) {}
+            }
+            /* If tx FIFO is not full then add data to the FIFO */
+            else {
+                /* if there is already data in the tsrb buffer write that byte */
+                int byte = tsrb_get_one(&uart_tx_rb[uart]);
+                if (byte >= 0) {
+                    send_byte(uart, byte);
+                    while (tsrb_add_one(&uart_tx_rb[uart], data[i]) < 0) {}
+                }
```

Comment:
Why can't we leave that part to the `irq_handler_tx`.
Having the tsrb consumed from both the ISR and user code (with interrupts enabled) sounds like a recipe for bugs and seems unnecessary.

Reply:
Because the `irq_handler_tx` is only triggered when the FIFO is below some level, and this level goes from 1/8 to 7/8 empty. It can also be triggered after the FIFO is completely empty by setting the EOT flag. But this still means that you need to somehow "bootstrap" the transmit process: If there is no data in the FIFO no "FIFO is below level" ISR will be triggered. If I write directly to the FIFO when its not full I can make sure there will be no cases of data being stuck in the tsrb buffer.

> tsrb consumed from both the ISR and user code (with interrupts enabled)

Hmm true there is some racy conditions potential there, let me think about it a bit...

Reply:
> Hmm true there is some racy conditions potential there, let me think about it a bit...

Guarding with `irq_disable()` `irq_restore` the write part should do it. But I want to see if I find a simpler solution.

Reply:
> Why can't we leave that part to the `irq_handler_tx`?
> Having the tsrb consumed from both the ISR and user code (with interrupts enabled) sounds like a recipe for bugs and seems unnecessary.

I couldn't really think of a much simpler workflow, do you have other ideas @benpicco?

Reply:
I guarded the pertinent section with `irq_disable/restore` BTW

Reply:
> If I write directly to the FIFO when its not full I can make sure there will be no cases of data being stuck in the tsrb buffer.

better add a comment about this then.

Reply:
I added a comment